File: src/sdk/python/ainari_sdk/ainari_sdk/instance.py
# Copyright 2022-2026 Tobias Anker <tobias.anker@kitsunemimi.moe>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from . import ainari_request
from .access_context import AccessContext


def create_instance(context: AccessContext,
                 name: str,
                 template: str) -> dict:
    path = "/v1alpha/instance"
    json_body = {
        "name": name,
        "template": template,
    }
    return ainari_request.send_post_request(context,
                                            context.hanami_address,
                                            path,
                                            json_body)


def get_instance(context: AccessContext,
              instance_uuid: str) -> dict:
    path = f"/v1alpha/instance/{instance_uuid}"
    return ainari_request.send_get_request(context,
                                           context.hanami_address,
                                           path,
                                           "")


def list_instances(context: AccessContext) -> dict:
    path = "/v1alpha/instance"
    return ainari_request.send_get_request(context,
                                           context.hanami_address,
                                           path,
                                           "")


def delete_instance(context: AccessContext,
                 instance_uuid: str):
    path = f"/v1alpha/instance/{instance_uuid}"
    ainari_request.send_delete_request(context,
                                       context.hanami_address,
                                       path,
                                       "")


def delete_all_instance(context: AccessContext):
    body = list_instances(context)["instances"]
    for entry in body:
        delete_instance(context, entry["uuid"])


def train(context: AccessContext,
          torii_port: int,
          instance_uuid: str,
          inputs: dict,
          outputs: dict):
    address = f"{context.torii_base_address}:{torii_port}"
    path = f"/v1alpha/instance/{instance_uuid}/train"
    json_body = {
        "inputs": inputs,
        "outputs": outputs,
    }
    ainari_request.send_put_request(context,
                                    address,
                                    path,
                                    json_body)


def request(context: AccessContext,
            torii_port: int,
            instance_uuid: str,
            inputs: dict,
            outputs: list):
    address = f"{context.torii_base_address}:{torii_port}"
    path = f"/v1alpha/instance/{instance_uuid}/request"
    json_body = {
        "inputs": inputs,
        "outputs": outputs,
    }
    return ainari_request.send_put_request(context,
                                           address,
                                           path,
                                           json_body)

File: src/sdk/go/ainari_sdk/instance.go
/**
 * @author      Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 * @copyright   Apache License Version 2.0
 *
 *      Copyright 2022-2026 Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 */

package ainari_sdk

import (
	"fmt"
)

func CreateInstance(context AccessContext, name, template string) (map[string]interface{}, error) {
	path := "v1alpha/instance"
	jsonBody := map[string]interface{}{
		"name":     name,
		"template": template,
	}
	return SendPost(context, context.HanamiAddress, path, jsonBody)
}

func GetInstance(context AccessContext, instanceUuid string) (map[string]interface{}, error) {
	path := fmt.Sprintf("v1alpha/instance/%s", instanceUuid)
	vars := map[string]interface{}{}
	return SendGet(context, context.HanamiAddress, path, vars)
}

func ListInstance(context AccessContext) (map[string]interface{}, error) {
	path := "v1alpha/instance"
	vars := map[string]interface{}{}
	return SendGet(context, context.HanamiAddress, path, vars)
}

func DeleteInstance(context AccessContext, instanceUuid string) (map[string]interface{}, error) {
	path := fmt.Sprintf("v1alpha/instance/%s", instanceUuid)
	vars := map[string]interface{}{}
	return SendDelete(context, context.HanamiAddress, path, vars)
}
